[DP-61] Adding a word panel

patelShas opened the pull request:
This adds a word panel to the side of the translation tab that can be opened up by clicking on a word.

![image](https://user-images.githubusercontent.com/92126390/153958074-19295c17-d7b1-4175-b8f9-d651b42d644c.png)

This Word panel remains on the side even as the main text is scrolled past or before, and a rectangle points out what word has been selected. The word panel has an "X" that will close the panel, but clicking on another word will change what word's details the panel is showing. 

![image](https://user-images.githubusercontent.com/92126390/153958590-648dead2-5280-454d-aa74-5891c8ae323a.png)

On a smaller screen or mobile device, the word panel will appear on a drawer from the right side. This drawer can be closed by either clicking the "X", or simply clicking off the drawer.

loafofpiecrust commented on website/src/word-panel.css.tsx:
```diff
@@ -0,0 +1,32 @@
+import sprinkles from "./sprinkles.css"
+
+export const wholePanel = style({
+    position: "sticky",
```
I never knew about this value for `position`! Looks like it has [pretty good browser support as long as you don't use it on table elements](https://caniuse.com/?search=sticky), but only if we don't do IE support. Looks like there's [a polyfill](https://github.com/wilddeer/stickyfill) if we do need to support IE. Great find!

loafofpiecrust commented on website/src/word-panel.css.tsx:
```diff
@@ -0,0 +1,32 @@
+    cursor: "pointer",
+    fontSize: "16px",
+    borderRadius: "0.25rem",
+    ":before":{
+        display: "inline-block",
+        margin: "4px",
+    },
```
I'm not certain this works? With [vanilla-extract](https://vanilla-extract.style/documentation/styling-api/) you normally need to use
```ts
selectors: {
  "&:before": {
    ...
  }
}
```

patelShas replied:
Yeah, I wasn't quite sure about how to write this section. I must have left that part out.

loafofpiecrust replied:
Just checked the docs again, and for simple pseudo-selectors you don't need to nest them under "selectors", so the code you wrote should work just fine.

loafofpiecrust commented on website/src/word-panel.css.tsx:
```diff
@@ -0,0 +1,32 @@
+        margin: "4px",
+    },
+    position: "absolute",
+    top: "5px",
+    right: "5px",
```
Use `vspace` and `hspace` for padding, margins, and absolute positioning.

loafofpiecrust commented on website/src/word-panel.css.tsx:
```diff
@@ -0,0 +1,32 @@
+    boxSizing: "border-box",
+    border: "transparent",
+    cursor: "pointer",
+    fontSize: "16px",
```
Use a relative [font size in `rem` units](https://www.sitepoint.com/understanding-and-using-rem-units-in-css/), for example this might just be `1.1rem` if you want bigger than normal body text or `0.9rem` if you want a bit smaller.

loafofpiecrust commented on website/src/word-panel.tsx:
```diff
@@ -0,0 +1,87 @@
+                        </div>
+                    </>
+                    ) : (
+                    (p.segment.audioTrack && (
```
You can use `x ? a : b` instead of `(x && a) || b` here.

loafofpiecrust commented on website/src/word-panel.tsx:
```diff
@@ -0,0 +1,87 @@
+                    <Button className={css.wordPanelButton} onClick={() => p.setContent(null)}>X</Button>
+
+                    <h2>{p.segment.source}</h2> 
+                    {p.segment.simplePhonetics ? (
```
Since there's like three different options here based on two conditions, I would extract this out into a flatter structure. Maybe like

```ts
let wordBreakdown = null
if (x) {
  wordBreakdown = ...
} else if (y) {
  wordBreakdown = ...
}

return (
  <div>
    {wordBreakdown}
  </div>
)
```

loafofpiecrust reviewed (commented):
Thanks for the full description and screenshots, super helpful!

- I'd like to see the details panel fill closer to the viewport height. 
- Let's also make the box around the currently selected word go around all of it's contents. Right now it's just around the first line, but I think surrounding the whole thing makes sense.
- Let's use an icon for the close button. I think we are for the `MorphemeDialog` component, so we should do something similar. Icon button with an `aria-label` that says "Close word details" or similar.
- It looks like, at least on mobile, that the fonts for the interlinear gloss aren't quite right. Make sure to include that styling, it really matters for the word analysis b/c that often has certain accents.
- Now that I'm looking at it, we should make sure we provide some guideposts especially while this panel lacks much info. So, would you mind changing the header to "Selected Word" and putting the Cherokee syllabary of the word below that?
- On mobile, the border shouldn't sit inside the sidebar. Either remove the border on mobile or remove the margin separating it from the edge.

patelShas commented on website/src/pages/documents/document.page.tsx:
```diff
@@ -132,6 +135,21 @@ const TranslationTab = ({ doc }: { doc: Document }) => {
   }
 
+  const dialog = useDialogState({ animated: true})
+  const [wordPanelContents, setwpContents] = useState<Dailp.FormFieldsFragment | null>(null)
```
Yeah, those names would be better fitting for the functions. I'll make the changes.

patelShas commented on website/src/pages/documents/document.page.tsx:
```diff
@@ -193,12 +248,14 @@ const DocumentContents = ({
   tagSet: TagSet
   openDetails: (morpheme: any) => void
   phoneticRepresentation: PhoneticRepresentation
+  wordPanelInfo: wordPanelDetails
```
I wasn't sure how else I could pass on the information on the Word Panel, since the word panel itself is on the translation tab, and the segments I want to get it passed to are in the DocumentContents in the translation tab. It seemed easier to pass it down in an object than to pass down each thing individually.

loafofpiecrust replied:
Yeah my feeling is if we are (almost) always passing these fields together and there are more than 2, then a sub-object makes sense. We can keep it for now at least.

patelShas commented on website/src/segment.tsx:
```diff
@@ -85,13 +89,22 @@ export const AnnotatedForm = (
-        <div className={css.syllabaryLayer} lang="chr">
+        
+        
+        <div 
+        className={wordCSS} 
+        lang="chr" 
+        onClick={() => p.wordPanelInfo.setCurrContents(p.segment)}>
```
Yeah, I wasn't aware of format on save, so I'll enable and use on all changes.

patelShas commented:
Update and made several changes to the word panel.

![image](https://user-images.githubusercontent.com/92126390/154331385-0cb348a9-2101-42c2-87f3-00ee3481ec05.png)

Notable changes:
- The word panel has been brought closer to the top
- The box for the selected word now circles the entire form, not just the word
- The close button now has an icon rather than just "X"
- "Selected Word:" added to the word panel

![image](https://user-images.githubusercontent.com/92126390/154331781-071b3db3-76e0-4371-9385-199f641251c2.png)

And on mobile:
- The word panel no longer has a box outline inside the drawer
- The font is the same as on desktop and the main page

patelShas commented:
![image](https://user-images.githubusercontent.com/92126390/154595719-4ebfedef-f379-401c-a0c0-717bb183b836.png)

The selected word is now more distinctive.

loafofpiecrust commented on website/src/pages/documents/document.page.tsx:
```diff
@@ -172,17 +215,31 @@ const TranslationTab = ({ doc }: { doc: Document }) => {
-          }}
-        />
-      </article>
+      <section className={css.contentContainer}>
```
I wouldn't use the `section` semantic element here because you are just grouping two things together, it's not really a thematic section of content. I would probably use a `div` here.

loafofpiecrust commented on website/src/word-panel.tsx:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    return (
+      <div className={css.wholePanel}>
+        <div className={css.wordPanelContent}>
```
So looks like you have the word panel nested in three `div`s. Two here, and one in `document.page.tsx`. Just for layout you almost never need three directly nested elements, so let's collapse the one in `document.page.tsx`.

loafofpiecrust commented on website/src/segment.css.ts:
```diff
@@ -63,7 +62,15 @@ export const wordGroup = style([
+  },
+])
+
+export const wordGroupSelection = styleVariants({
```
Not sure where in the styling this happens, but words get all messed up when you select them. [I'm looking here](https://deploy-preview-147--dailp.netlify.app/documents/dd6); when I select one word, then select another they shift around in the document. We really can't have that, we need the words to sit still when we (de)select them.

Looks like it might be this: You can't change the padding or margin between the selected and unselected states because that will change the layout flow.

loafofpiecrust commented on website/src/segment.tsx:
```diff
@@ -85,13 +89,21 @@ export const AnnotatedForm = (
+        <div
+          className={css.syllabaryLayer}
+          lang="chr"
+          onClick={() => p.wordPanelDetails.setCurrContents(p.segment)}
```
As I play with it, I'd like users to be able to highlight the text of a word without changing their selection state. So, can we only select the word if you click on the information icon in the top-right corner of each word? That's how the designs are envisioned.

For this, you should also change the cursor style for the info icon to be `select`. And please add what currently shows as a tooltip when you hover on the info icon to the word details panel, below the multi-line translation. And then get rid of the tooltip.

patelShas replied:
That makes sense, but I'd like to clarify what you mean about the cursor style for the info icon. As far as I can tell, select is not a single cursor style but a category where there are multiple. The most similar I think is `text` which shows the cursor as an I-beam. Even still, I don't quite get why we would use this, as it usually shows text selection, and the info icon is simply something we click on. 

Do you mean having this text selection icon for the whole word except the info icon, or some other cursor style?

loafofpiecrust commented on website/src/word-panel.css.tsx:
```diff
@@ -0,0 +1,39 @@
+import { style } from "@vanilla-extract/css"
+import sprinkles, { hspace, mediaQueries, theme, vspace } from "./sprinkles.css"
+
+export const wholePanel = style({
```
I think I mentioned this before, but we do want the panel to fill to the bottom of the screen. This way, it doesn't change height if a word has more information than another.

loafofpiecrust commented on website/src/word-panel.css.tsx:
```diff
@@ -0,0 +1,39 @@
+  top: 125,
+})
+
+export const wordPanelButton = style({
```
You can look at the close button for the `MorphemeDialog` which already has some spacing styles & clears out the background so it's just the icon. I think that styling will look better here.

loafofpiecrust commented on website/src/word-panel.tsx:
```diff
@@ -0,0 +1,95 @@
+          >
+            <MdClose size={32} />
+          </Button>
+          <h2>Selected word:</h2>
+          <h1>{p.segment.source}</h1>
```
Swap the header types here, use an `h1` for "Selected Word" and an `h2` for the word source text, since that follows "main header" and "subheader".

patelShas commented:
![image](https://user-images.githubusercontent.com/92126390/155397338-42e59528-61d3-4c43-85bc-4834b0b84c2f.png)

Added the latest changes:

- Words now all have info icons. The word panel is now opened by clicking this info icon, not the word's details.
- Word commentary is now displayed on the Word panel after other details.
- Padding is adjusted so that selecting different words doesn't change how they are laid out on the page.
- The word panel is extended to fill out more of the page.
- The heading and close buttons have both been changed in appearance.

patelShas commented:
Just changed height of the word panel to scale a bit more with the screen's actual size.

loafofpiecrust commented:
It looks like you're assuming words are uniquely identified by their syllabary layer, which is not true. Make sure to check the word index instead, so that you don't get duplicate selections like this:

<img width="1000" alt="Screen Shot 2022-02-28 at 11 10 10 AM" src="https://user-images.githubusercontent.com/3489148/156044211-fdac2975-a529-43ca-8523-f69cb744b94b.png">

patelShas commented:
- Fixed up the fonts on the word panel to be more correct.
- Fixed an issue where if a word appeared multiple times in the document, all occasions would be highlighted when the word was selected. Now, only the one clicked will be highlighted.

patelShas commented:
Fixed another issue where on smaller screen or mobile devices (where the word panel appeared as a drawer from the left) the selected word would still be highlighted after the word panel is closed. Closing the word panel now removes the highlight.

patelShas commented:
- Made it so the word panel updates immediately when experience level/analysis is changed
- Allowed detailed morpheme dialog to appear without closing the word panel
- The word panel will close when switching to story mode
